### PR TITLE
feat(StopPageDepartures): fetch predictions at top level

### PR DESCRIPTION
**Pull request by thecristen**
#### Summary of changes

No ticket. It moves our call to predictions higher up in the tree, making one websocket connection for all the predictions, instead of one per route/direction.

---

#### General checks
* [ ] **Are the changes organized into self-contained commits with descriptive and well-formatted commit messages?** This is a good practice that can facilitate easier reviews.
* [ ] **Testing.** Do the changes include relevant passing updates to tests? This includes updating screenshots. Preferably tests are run locally to verify that there are no test failures created by these changes, before opening a PR.
* [ ] **Tech debt.** Have you checked for tech debt you can address in the area you're working in? This can be a good time to address small issues, or create Asana tickets for larger issues.

#### New UI, or substantial UI changes
* [ ] **Cross-browser compatibility** is less of an issue now that we're no longer supporting IE, but changes still need to work as expected in Safari, Chrome, and Firefox.
* [ ] **Are interactive elements accessible?** This includes at minimum having [relevant keyboard interactions](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#keyboard-interaction) and [visible focus](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#visible-focus), but can also include verification with screen reader testing.
* [ ] **Other accessibility checks** such as sufficient [color constrast](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#color-contrast), or whether the layout holds up at 200% [zoom level](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#visible-focus).

#### New endpoints, or non-trivial changes to current endpoints
* [ ] **Have we load-tested** any new pages or internal API endpoints that will receive significant traffic? [See load testing docs](./../apps/site/load_tests/README.md)
* [ ] **If this change involves routes,** does it work correctly with pertinent "unusual" routes such as the combined Green Line, Silver Line, Foxboro commuter rail, and single-direction bus routes like the 170?

### CODE REVIEW

File: apps/site/assets/ts/stop/components/StopPageDepartures.tsx [hunk 1]
```diff
@@ -14,7 +17,7 @@ interface StopPageDeparturesProps {
   onClick: (
     route: Route,
     directionId: DirectionId,
-    departures: ScheduleWithTimestamp[] | null | undefined
+    departures: DepartureInfo[] | null | undefined
```
Comment (kotva006) [inline apps/site/assets/ts/stop/components/StopPageDepartures.tsx:20]: So the one downside to this approach is that the predictions will not update on the realtime list.  The value of this property `departures` is set when the user clicks on a row.  When the `usePredictionsChannel` updates with new predictions it never gets a chance to pass them to the realtime list.  The user is on a different component, and this predictions values are never passed to the list directly.

File: apps/site/assets/ts/hooks/__tests__/usePredictionsChannelTest.tsx [hunk 2]
```diff
@@ -68,7 +70,11 @@ describe("usePredictionsChannel hook", () => {
 
   test("gets and outputs data", () => {
     const { result } = renderHook(() =>
-      usePredictionsChannel("Purple", "place-somewhere", 0)
+      usePredictionsChannel({
```
Comment (i0sea) [inline apps/site/assets/ts/hooks/__tests__/usePredictionsChannelTest.tsx:73]: nitpick: can a test be added that only uses `stopId`?

File: apps/site/assets/ts/stop/__tests__/DepartureListTest.tsx [hunk 3]
```diff
@@ -42,40 +42,26 @@ const schedules = [
...
+  }
+] as PredictionWithTimestamp[];
 
-  afterAll(() => {
-    jest.restoreAllMocks();
-  });
+const departures = mergeIntoDepartureInfo(schedules, predictions);
 
+describe("DepartureList", () => {
   it("should render a schedule when no predictions available", () => {
```
Comment (i0sea) [inline apps/site/assets/ts/stop/__tests__/DepartureListTest.tsx:59]: add test for rendering prediction if there are no schedules

**Review by i0sea — CHANGES_REQUESTED** [review]

File: apps/site/assets/ts/hooks/usePredictionsChannel.ts [hunk 4]
```diff
@@ -66,45 +65,46 @@ export const parsePrediction = (
...
-  );
-};
+interface PredictionsChannelArgs {
+  routeId?: string;
+  stopId?: string;
+  directionId?: 0 | 1;
+}
+
+function channelFromArgs(channelArgs: PredictionsChannelArgs): string {
+  const keysWithValues = Object.entries({
```
[the remark this answers is not in the document]
Comment (thecristen) [reply apps/site/assets/ts/hooks/usePredictionsChannel.ts:75]: It does! Nice catch, using an empty string is probably suboptimal... will update to not try to join channels..

**Review by i0sea — APPROVED** [review]
🚀

**Pull request merged**